### PR TITLE
migrate aws configurator to sdk v2

### PR DESCRIPTION
**Pull request by GavinFrazar**
Part of:
- https://github.com/gravitational/teleport/issues/14142

This PR migrates the aws configurator to aws sdk v2.

### CODE REVIEW

File: lib/cloud/aws/identity.go [hunk 1]
```diff
@@ -23,6 +23,8 @@ import (
 	"fmt"
 	"strings"
 
+	awsV2 "github.com/aws/aws-sdk-go-v2/aws"
```
Comment (zmb3) [inline lib/cloud/aws/identity.go:26]:
```suggestion
	awsv2 "github.com/aws/aws-sdk-go-v2/aws"
```
Try to keep package names (and their aliases) all lowercase

File: lib/cloud/aws/policy.go [hunk 2]
```diff
@@ -25,9 +25,9 @@ import (
 	"slices"
 	"sort"
 
-	"github.com/aws/aws-sdk-go/aws"
-	"github.com/aws/aws-sdk-go/service/iam"
-	"github.com/aws/aws-sdk-go/service/iam/iamiface"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/iam"
+	iamTypes "github.com/aws/aws-sdk-go-v2/service/iam/types"
```
Comment (zmb3) [inline lib/cloud/aws/policy.go:30]:
```suggestion
	iamtypes "github.com/aws/aws-sdk-go-v2/service/iam/types"
```

File: lib/cloud/aws/policy.go [hunk 3]
```diff
@@ -404,13 +400,23 @@ type policies struct {
 	partitionID string
 	// accountID current AWS account ID.
 	accountID string
-	// iamClient already initialized IAM client.
-	iamClient iamiface.IAMAPI
+	// iamClient is an already initialized IAM client.
+	iamClient iamClient
+}
+
+type iamClient interface {
```
Comment (zmb3) [inline lib/cloud/aws/policy.go:407]: I'd export this. Otherwise you have an exported constructor func that requires as input an unexported interface type.

File: lib/cloud/aws/policy_test.go [hunk 4]
```diff
@@ -575,7 +574,7 @@ func TestPolicyEnsureStatements(t *testing.T) {
 	}, policy)
 }
 
-func TestRetrievePolicy(t *testing.T) {
+func Test_getPolicyVersions(t *testing.T) {
```
Comment (zmb3) [inline lib/cloud/aws/policy_test.go:577]:
```suggestion
func TestGetPolicyVersions(t *testing.T) {
```

File: lib/configurators/aws/aws.go [hunk 5]
```diff
@@ -327,58 +333,51 @@ func (c *ConfiguratorConfig) CheckAndSetDefaults() error {
...
+
+		if c.ssmClients == nil {
+			c.ssmClients = make(map[string]ssmClient)
 			for _, matcher := range c.ServiceConfig.Discovery.AWSMatchers {
 				if !slices.Contains(matcher.Types, types.AWSMatcherEC2) {
 					continue
 				}
 				for _, region := range matcher.Regions {
-					if _, ok := c.AWSSSMClients[region]; ok {
+					region := region
```
Comment (zmb3) [inline lib/configurators/aws/aws.go:363]: Why is this necessary?

File: lib/configurators/aws/aws.go [hunk 6]
```diff
@@ -327,58 +333,51 @@ func (c *ConfiguratorConfig) CheckAndSetDefaults() error {
...
-			c.AWSSession, err = awssession.NewSessionWithOptions(awssession.Options{
-				SharedConfigState: awssession.SharedConfigEnable,
-				Config: aws.Config{
-					UseFIPSEndpoint: useFIPSEndpoint,
-				},
-			})
-			if err != nil {
-				return trace.Wrap(err)
-			}
+		awsConfig, err := config.LoadDefaultConfig(ctx,
```
Comment (greedy52) [inline lib/configurators/aws/aws.go:336]: will the unit test fails if host does not have aws credentials?

File: lib/configurators/aws/aws_test.go [hunk 7]
```diff
@@ -1388,16 +1385,15 @@ func TestAWSDocumentConfigurator(t *testing.T) {
 	require.NoError(t, config.ApplyFileConfig(fileConfig, serviceConfig))
 
 	config := ConfiguratorConfig{
-		AWSSession:   &awssession.Session{},
-		AWSIAMClient: &iamMock{},
-		AWSSTSClient: &STSMock{ARN: "arn:aws:iam::1234567:role/example-role"},
-		AWSSSMClients: map[string]ssmiface.SSMAPI{
-			"eu-central-1": &SSMMock{
+		iamClient: &iamMock{},
+		stsClient: &stsMock{ARN: "arn:aws:iam::1234567:role/example-role"},
```
Comment (greedy52) [inline lib/configurators/aws/aws_test.go:1389]: nit: IMHO, ideally tests should mock an identity instead of `stsClient`. otherwise every place calls `GetIdentityWithClientV2` will need to mock sts, which is kinda indirect. but i think we do this all the places anyway.

**Review by greedy52 — APPROVED** [review]

File: lib/cloud/aws/policy.go [hunk 8]
```diff
@@ -404,13 +400,23 @@ type policies struct {
...
+	AttachRolePolicy(ctx context.Context, params *iam.AttachRolePolicyInput, optFns ...func(*iam.Options)) (*iam.AttachRolePolicyOutput, error)
+	AttachUserPolicy(ctx context.Context, params *iam.AttachUserPolicyInput, optFns ...func(*iam.Options)) (*iam.AttachUserPolicyOutput, error)
+	CreatePolicy(ctx context.Context, params *iam.CreatePolicyInput, optFns ...func(*iam.Options)) (*iam.CreatePolicyOutput, error)
+	CreatePolicyVersion(ctx context.Context, params *iam.CreatePolicyVersionInput, optFns ...func(*iam.Options)) (*iam.CreatePolicyVersionOutput, error)
+	DeletePolicyVersion(ctx context.Context, params *iam.DeletePolicyVersionInput, optFns ...func(*iam.Options)) (*iam.DeletePolicyVersionOutput, error)
+	GetPolicy(ctx context.Context, params *iam.GetPolicyInput, optFns ...func(*iam.Options)) (*iam.GetPolicyOutput, error)
+	ListPolicyVersions(ctx context.Context, params *iam.ListPolicyVersionsInput, optFns ...func(*iam.Options)) (*iam.ListPolicyVersionsOutput, error)
 }
```
Comment (espadolini) [inline lib/cloud/aws/policy.go:415]: Just to clarify what this interface is supposed to abstract over and give an example.

```suggestion
}

var _ iamClient = (*iam.Client)(nil)
```

File: lib/cloud/aws/identity.go [hunk 9]
```diff
@@ -106,6 +108,10 @@ func (i identityBase) String() string {
 	return i.arn.String()
 }
 
+type callerIdentityGetter interface {
+	GetCallerIdentity(ctx context.Context, params *stsV2.GetCallerIdentityInput, optFns ...func(*stsV2.Options)) (*stsV2.GetCallerIdentityOutput, error)
+}
+
 // GetIdentityWithClient determines AWS identity of this Teleport process
 // using the provided STS API client.
 func GetIdentityWithClient(ctx context.Context, stsClient stsiface.STSAPI) (Identity, error) {
```
Comment (espadolini) [inline lib/cloud/aws/identity.go:117]: How long would it take to get rid of STS v1 in its entirety?

**Review by espadolini — APPROVED** [review]
LGTM other than the package naming that @zmb3 mentioned.

File: lib/configurators/aws/aws.go (hunk 5 above)
Comment (GavinFrazar) [reply lib/configurators/aws/aws.go:363]: region is captured by the closure below (sdk v2 config uses the closure, v1 did not).
With go >=1.22 it works without reassignment, but on older go versions I think it would be bugged. branch/v14 is still on 1.21. I have no plans to backport this to v14 but I figured it didn't hurt to guard against it

File: lib/configurators/aws/aws_test.go (hunk 7 above)
Comment (GavinFrazar) [reply lib/configurators/aws/aws_test.go:1389]: I just converted what was here already, but I think mocking sts makes the test closer to reality than mocking identity directly

File: lib/configurators/aws/aws.go (hunk 5 above)
Comment (zmb3) [reply lib/configurators/aws/aws.go:363]: But `o.Region = region` is a string assignment, which makes a copy.

Are you sure this is necessary?

File: lib/cloud/aws/identity.go (hunk 9 above)
Comment (GavinFrazar) [reply lib/cloud/aws/identity.go:117]: I'm not sure I only touched this package because it was being used by lib/configurators/aws

Probably would not take very long, but I'd rather not include it in this PR

File: lib/configurators/aws/aws.go (hunk 6 above)
Comment (GavinFrazar) [reply lib/configurators/aws/aws.go:336]: ah good point, it does not. I'll fix that.

Comment (GavinFrazar) [reply lib/configurators/aws/aws.go]: done https://github.com/gravitational/teleport/pull/46367/commits/24b9ab917818ba69a533d964980925132e85afa1

File: lib/configurators/aws/aws.go (hunk 5 above)
Comment (GavinFrazar) [reply lib/configurators/aws/aws.go:368]: evaluation is delayed until the config is actually used, but actually I see that we're applying the option within the loop, so you're right. I'll remove that reassignment then.
